### PR TITLE
Add imagePullSecrets to CephCluster CRD

### LINKED ISSUE #6673: Mon pods not able to use image pull secrets
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
When a private docker registry is used and an image pull secret is specified in the chart, the mon pods fail to pull the image due to authentication issues. The reason is that mon and mon-canary pods do not use a custom service account like osd and other components but rather rely on the default service account which is not aware of the custom image pull secrets.

**Expected behavior:**
A service account for mon pods should exist and it should define image pull secrets like other service account in rook-ceph.

**How to reproduce it (minimal and precise):**
* Install rook-ceph and populate `imagePullSecrets` in values.yaml
* Define a CephCluster resource with a ceph/ceph image from a private docker registry
* mon pods will fail schedule with error _Failed to pull image ..._, with an authentication error that depends on the docker registry.

**Environment**:
Any kubernetes cluster installation

### PR DESCRIPTION
**Pull request by ddl-kevin**
**Description of your changes:**

Introduce optional field on `CephCluster` CRD to specify `imagePullSecrets` which should be used on all cluster Pods. This is a solution which does not rely on customized RBAC roles managed by the Helm charts.

Note that it did not make sense to add image pull secrets to CSI or Discover Pods.

**Which issue is resolved by this Pull Request:**
Resolves #6673

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by ddl-kevin** [comment]
The build failures look to be flaky. I'm able to reproduce sporadically when building with parallelism (`-j4`).

**Review by travisn — CHANGES_REQUESTED** [review]
Have you looked at [adding the image pull secrets to the service accounts](https://rook.io/docs/rook/latest/Getting-Started/Prerequisites/authenticated-registry/)? With that approach, we shouldn't need to add the pull secrets to all the pods individually. 

If the setting is in the cluster CR it does also have the limitation of not applying to to the csi and discovery pods as you mentioned as well.

#6673 is about using a specific service account for all the rook pods instead of the default service account, which would just change which service account(s) need the image pull secrets.

**Comment by ddl-kevin** [comment]
> Have you looked at [adding the image pull secrets to the service accounts](https://rook.io/docs/rook/latest/Getting-Started/Prerequisites/authenticated-registry/)? With that approach, we shouldn't need to add the pull secrets to all the pods individually.
> 
> If the setting is in the cluster CR it does also have the limitation of not applying to to the csi and discovery pods as you mentioned as well.
> 
> #6673 is about using a specific service account for all the rook pods instead of the default service account, which would just change which service account(s) need the image pull secrets.

I have seen the alternative suggested, which requires adding the secret to the `default` service account. This seems like a very hacky solution, and it does not seem like the issue blocking the use of an empty service account for `mon` pods will be fixed soon. 

It is currently a complete blocker if adding the pull secret to the `default` SA is not permitted. This adds an escape valve, and is (IMO) a more correct solution, rather than relying on logic outside of the operator (the Helm chart).

If we want to support this for CSI pods, I would support that, but I believe that should be configured via operator arguments and not the Cluster CRD.

**Comment by travisn** [comment]
To move ahead with the service account solution, instead of waiting on the operator sdk dependency mentioned in #6673, I'd suggest we bring back #7406 and add a workaround as seen in #10142. This will allow us to avoid the default service account, and also avoid adding a new setting to the cluster CR. @parth-gr Want to take a look at this?

**Comment by ddl-kevin** [comment]
> To move ahead with the service account solution, instead of waiting on the operator sdk dependency mentioned in #6673, I'd suggest we bring back #7406 and add a workaround as seen in #10142. This will allow us to avoid the default service account, and also avoid adding a new setting to the cluster CR. @parth-gr Want to take a look at this?

Can you explain the desire to keep this configuration out of the CR? There are a number of very similar fields in the CR for setting resource requests, labels, annotations, and many other Pod attributes. I think this is generally best practice for Operators.

**Comment by travisn** [comment]
> > To move ahead with the service account solution, instead of waiting on the operator sdk dependency mentioned in #6673, I'd suggest we bring back #7406 and add a workaround as seen in #10142. This will allow us to avoid the default service account, and also avoid adding a new setting to the cluster CR. @parth-gr Want to take a look at this?
> 
> Can you explain the desire to keep this configuration out of the CR? There are a number of very similar fields in the CR for setting resource requests, labels, annotations, and many other Pod attributes. I think this is generally best practice for Operators.

My thoughts:
- If the image secrets can be applied to service accounts, there is no need to expose it in the CRDs or apply to individual pod specs, keeping the operator simpler
- If it's added to the CephCluster, we really should add it to other places for consistency:
   - operator.yaml for the csi driver and discovery pods
   - Each of the CRDs that start a pod, such as CephFilesystem (mds), CephObjectStore (rgw), CephRBDMirror (rbd mirror), etc

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 1]
```diff
@@ -220,6 +220,11 @@ type ClusterSpec struct {
 	// +optional
 	// +nullable
 	LogCollector LogCollectorSpec `json:"logCollector,omitempty"`
+
+	// Image pull secrets to apply to all Ceph Pods
+	// +optional
+	// +nullable
+	ImagePullSecrets []v1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:227]: Similar to #10966, how about we move this setting under the `CephVersionSpec`? It is really related to the pulling of the image which is defined in that struct, and would follow the same pattern as adding the ImagePullPolicy in #10966.

**Review by travisn — CHANGES_REQUESTED** [review]
To add to my previous comments... While I like the service account approach, this feature would be simple enough to support as you've also implemented it and I don't have a good reason to say we shouldn't do this. This may even be simpler to use than modifying the service accounts. So let's proceed with this PR, and no need to raise the priority of implementing the service accounts mentioned [here](https://github.com/rook/rook/pull/10956#issuecomment-1244480517).

For the CSI driver pull secrets, I believe you could also follow the pattern in #10966 for the csi driver settings.

**Comment by ddl-kevin** [comment]
Awesome, sounds good!

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 1 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:227]: @ddl-kevin Planning on making this change so we can get this in? Thanks!

Comment (ddl-kevin) [reply pkg/apis/ceph.rook.io/v1/types.go:227]: Hey, sorry about this. We've gone in a different direction for our use case. I will see if I have some cycles this weekend to get this across the line.

**Pull request closed without merging**